### DISCUSSION TITLE
Binding Issues in Custom MAUI Component

### DISCUSSION THREAD

**Discussion by GebekovAS** [dotnet/maui#23569] [Q&A] [answered]
Hi everyone,

I'm facing an issue with data binding in my custom MAUI component. I have a custom control with field (`Text`), and I want to bind their value to propertie in my ViewModel. However, the value in the ViewModel do not update when I change the field in the UI.

Here is the relevant part of my custom component's XAML:

```xml
<Entry Text="{Binding Text, Source={x:Reference this}}" />
```

And here is the code-behind for the custom component:

```csharp
public partial class MyCustomControlView : ContentView
{
    public static readonly BindableProperty TextProperty =
        BindableProperty.Create(nameof(Text), typeof(string), typeof(MyCustomControlView), default(string), BindingMode.TwoWay);

    public string Text
    {
        get => (string)GetValue(TextProperty);
        set => SetValue(TextProperty, value);
    }

    // Other methods and event handlers...
}
```

And here is how I use the custom component in my main view:

```xml
<avControls:MyCustomControlView Text="{Binding ViewModelTextField}" />
```

Despite setting the binding mode to `TwoWay`, changes made in the UI do not reflect in the ViewModel.
What could be causing this issue? Any help would be greatly appreciated!

---

Thank you in advance for your assistance!

**Comment by acaliaro** [comment]
maybe ViewModelTextField is not observable?

**Reply by GebekovAS** [reply]
View model code
```cs
public class MyViewModel : ObservableObject, INotifyPropertyChanged
{
    private string _viewModelTextField;
    public string ViewModelTextField
    {
        get => _viewModelTextField;
        set => SetProperty(ref _viewModelTextField, value);
    }
    ...
```
Binding for primitive controls (entry) for this model works. Not work only the custom control

**Reply by acaliaro** [reply]
here you can user [ObservableProperty] instead of have get and set

**Reply by GebekovAS** [reply]
Thank you, your advice was quite helpful and made my code more elegant. The improvements work well for other fields, but I'm still facing issues with binding when it comes to the custom component.

**Comment by acaliaro** [comment, marked as answer]
![image](https://github.com/user-attachments/assets/0d7ef5b3-aa1f-43a8-9ad9-9d6b51decb22)
I think you have to set the bindingcontext in you customcontrol, not the source

**Reply by GebekovAS** [reply]
Thank you very much, this really helped. Have a good day!

**Reply by vkbishnoi** [reply]
I was doing BindingContext = this, in the custom control xaml.cs file in the constructor. It didn't worked for me. What worked is using x:Reference as suggested in this post.
